[FEAT] 이번달 챙길사람 기능

stopstone opened the pull request:
## 작업 내용
홈 화면에서 이번달 챙길사람 화면으로 이동합니다.
이번달 챙길사람 UI 구성
챙김 기록하기 기능

## 확인 방법
- feature/mothlyreminderall
- FriendRepository - fetchMonthlyCompleteFriends()
- FriendService - fetchMonthlyCompleteFriends()

## 참고 사항
- 만약 이번달 챙길 사람은 없고, 챙김 완료한 사람이 있는 경우는 
"이번달은 챙길 사람이 없네요"를 표기해야하는지
챙김 완료만 보여주는지 확인 후 수정
현재는 이번달 챙김 리스트와 챙김 완료가 모두 비어야 빈 화면을 표시합니다.

- 챙김 완료시 표시되는 날짜는 
챙긴 날짜를 표시하는 것인지, 
기념일을 표시하는 것인지 확인 후 수정
현재는 기념일을 표시합니다

## 관련 이슈
- close #56

stopstone committed:
feat: 이번달 챙길사람 전체보기 화면 네비게이션 추가

- 홈 화면의 '이달의 리마인더' 섹션에 '전체보기' 버튼을 추가하고, 클릭 시 해당 화면으로 이동하도록 구현했습니다.
- `ic_front_8` 벡터 드로어블을 추가하여 '전체보기' 버튼 아이콘으로 사용했습니다.
- 전체보기 화면으로 이동하는 `navigateToMonthlyReminderAll` 네비게이션 로직과 `MonthlyReminderAllScreen` 컴포저블을 추가했습니다.

stopstone committed:
feat: 월별 리마인더 친구 카드 컴포저블 추가

- 월별 전체 리마인더 화면에서 사용될 `MonthlyReminderFriendCard` 컴포저블을 추가했습니다.
- 이 카드는 친구의 이름, 리마인더 내용, D-day, 그리고 '챙김 기록하기' 버튼으로 구성됩니다.

stopstone committed:
feat: 월별 리마인더 화면 빈 상태 컴포저블 추가

- 월별 리마인더에 표시할 내용이 없을 때 보여줄 `MonthlyReminderEmpty` 컴포저블을 추가했습니다.
- 해당 컴포저블은 안내 이미지와 "이번달은 챙길 사람이 없네요."라는 텍스트로 구성됩니다.

stopstone committed:
feat: 월별 리마인더 완료 목록 아이템 컴포저블 추가

- 월별 리마인더의 완료된 항목을 표시하는 `MonthlyReminderComplete` 컴포저블을 새로 추가했습니다.
- 이 컴포저블은 아이콘, 이름, 날짜를 포함하며, 완료된 상태의 UI를 나타냅니다.

stopstone committed:
refactor: 월별 리마인더 전체보기 화면 파일 이동 및 ViewModel 연결

stopstone committed:
feat: 이번달 챙길 사람 전체보기 화면 구현

- UI 상태(로딩, 데이터 있음, 데이터 없음)에 따라 각각 다른 화면을 표시합니다.
- 뒤로가기 버튼을 통해 이전 화면으로 이동할 수 있습니다.

stopstone committed:
feat: 이달의 챙김 친구 카드 UI 모델 적용

stopstone committed:
feat: iewModel 및 UI State 구현

- `MonthlyReminderAllViewModel`을 추가하여 월별 리마인더 목록을 가져오고 UI 상태를 관리합니다.
- API 응답을 `MonthlyReminderUIModel`로 변환하여 UI에 필요한 데이터를 제공합니다.
- `MonthlyReminderAllUIState` sealed interface를 `Loading`, `Empty`, `Success` 상태로 정의하여 UI 상태를 관리하도록 수정했습니다.

stopstone committed:
feat: 이달의 챙김 완료 친구 목록 조회 API 추가

stopstone committed:
chore: MonthlyReminderFriendCard 패딩 삭제

stopstone committed:
feat: 챙김 완료 목록 아이템 UI 구현

- `MonthlyReminderComplete` 컴포저블이 `MonthlyReminderUIModel` 데이터를 받아 실제 데이터를 표시하도록 수정했습니다.
- 날짜 형식을 'yy.MM.dd'로 포맷하는 `formatDate` 함수를 추가했습니다.
- 이름이 길어질 경우 말줄임표(...)로 처리되도록 `TextOverflow.Ellipsis` 속성을 적용했습니다.

stopstone committed:
feat: 챙김 완료 목록 추가

stopstone committed:
feat: 연락 기록 성공 다이얼로그 추가

- 연락을 기록했을 때 표시되는 성공 다이얼로그 UI를 구현했습니다.
- 다이얼로그는 1.5초 후 자동으로 사라집니다.

stopstone committed:
refactor: 챙김 친구 완료 시 UI 즉시 업데이트

- 월별 리마인더 목록에서 완료 처리한 친구를 연락 완료 목록으로 즉시 이동시켜 UI에 반영하도록 수정했습니다.
- 기존의 네트워크 통신을 통한 목록 갱신 방식 대신, 로컬에서 목록을 직접 업데이트하여 불필요한 API 호출을 제거하고 응답성을 개선했습니다.

stopstone committed:
refactor: 월별 리마인더 없음 화면 UI 수정

stopstone committed:
refactor: 이번달 챙길 사람 화면 문자열 리소스 적용

stopstone committed:
refactor: 이번달 챙길 사람 화면 챙김 완료 상단 패딩 수정

- 불필요한 Spacer를 제거하여 레이아웃을 개선했습니다.
- LazyColumn의 컨텐츠 패딩을 적용하여 상하단 여백을 조정했습니다.

stopstone committed:
refactor: 연락 완료 여부 확인 로직을 uiState로 관리

stopstone committed:
refactor: 이번달 챙길 사람 화면 간격 추가

- 월별 리마인더 전체보기 화면의 카드 사이에 16dp의 간격을 추가했습니다.

stopstone committed:
feat: 이번달 챙길 사람 리스트 정렬

- 월별 리마인더 목록을 `nextContactAt`(다음 연락일) 기준으로 오름차순 정렬하여 표시하도록 수정했습니다.

stopstone committed:
refactor: 월별 리마인더 데이터 로딩 로직 개선

- `combine`을 사용하여 `fetchMonthlyFriends`와 `fetchMonthlyCompleteFriends` API 호출을 병렬로 처리하도록 수정했습니다.
- 두 API의 응답을 모두 받은 후에 한 번에 UI 상태를 업데이트하여 데이터 로딩 중 발생할 수 있는 화면 깜빡임 현상을 개선했습니다.
- API 응답을 함께 관리하기 위한 `MonthlyReminderCombinedData` data class를 추가했습니다.

stopstone committed:
refactor: 홈 화면 "전체 보기" 텍스트를 문자열 리소스로 분리

stopstone committed:
refactor: 에러 처리 방식 개선

stopstone committed:
refactor: Flow 에러 처리 로직 공통화

- 반복적으로 사용되는 Flow의 `catch` 블록 내 에러 처리 로직을 `handleError` 확장 함수로 분리하여 공통화했습니다.
- `MonthlyReminderAllViewModel` 내 `fetchMonthlyReminders`와 `onRecordFriendShip` 함수의 에러 처리 로직에 `handleError`를 적용하여 코드 중복을 제거하고 가독성을 개선했습니다.

stopstone committed:
refactor: 기록 완료 다이얼로그 색상 수정

- `RecordSuccessDialog`에서 사용하던 하드코딩된 색상 값을 `NearTheme`의 색상으로 변경했습니다.
- `NearColor`에 `BLACK_222222` 색상을 추가했습니다.

stopstone committed:
chore: MonthlyReminderFriendCard.kt 코드 스타일 수정

stopstone committed:
refactor: 이번달 챙길사람 하단 패딩 수정

stopstone committed:
chore: "D-day"를 "D-DAY"로 수정

- 연락 주기까지 남은 날짜를 표시하는 문자열을 "D-day"에서 "D-DAY"로 변경했습니다.

stopstone committed:
feat: 연락일이 오늘일 경우 D-day 스타일 적용 및 중복 데이터 제거

- `MonthlyReminderUIModel`에 연락일이 오늘인지 여부를 나타내는 `isToday` 프로퍼티를 추가했습니다.
- `isToday` 값에 따라 D-day 텍스트 스타일(파란색, 굵은 글씨)이 적용되도록 `MonthlyReminderFriendCard`를 수정했습니다.
- 월간 리마인더 목록에서 중복된 친구 데이터가 표시되지 않도록 `distinctBy`를 사용하여 필터링 로직을 추가했습니다.

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/mothlyreminderall/MonthlyReminderAllScreen.kt:
```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+@Preview
+@Composable
+private fun MonthlyReminderAllScreenEmptyPreview() {
+    NearTheme {
+        MonthlyReminderAllScreen(
+            uiState = MonthlyReminderAllUIState.Empty,
+        )
+    }
+}
+
+@Preview
+@Composable
+private fun MonthlyReminderAllScreenLoadingPreview() {
+    NearTheme {
+        MonthlyReminderAllScreen(
+            uiState = MonthlyReminderAllUIState.Loading,
+        )
+    }
+}
```
상세한 프리뷰 좋네요👍
[PreviewParameter](https://developer.android.com/develop/ui/compose/tooling/previews?hl=ko#preview-data)를 통해 하나의 프리뷰 주석으로 해결할 수도 있습니다!

괜찮으시다면 한 번 적용해보아도 좋을 것 같습니다!

stopstone replied:
오..! PreviewParameter라는 것이 있군요! 덕분에 알았습니다
확인해보고 적용해보겠습니다!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/mothlyreminderall/MonthlyReminderAllViewModel.kt:
```diff
@@ -0,0 +1,133 @@
+        val uiState: StateFlow<MonthlyReminderAllUIState> = _uiState.asStateFlow()
+
+        private val _monthlyReminders = MutableStateFlow<List<MonthlyReminderUIModel>>(emptyList())
+        private val _completedReminders = MutableStateFlow<List<MonthlyReminderUIModel>>(emptyList())
```
내부 데이터 캐싱을 위해 사용되나 보네요! 외부 노출이 안된다면
언더스코어는 제외해도 될 것 같습니다!

stopstone replied:
내부에서 사용되는 것들은 언더스코어로 다 작성을 하였는데
외부에서는 알 필요가 없는 프로퍼티는 생략하는 것이 좋겠네요 감사합니다!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/mothlyreminderall/MonthlyReminderAllScreen.kt:
```diff
@@ -0,0 +1,224 @@
+                        onShowErrorSnackBar(event.throwable)
+                    }
+
+                    is MonthlyReminderAllUIEvent.RecordFriendShipSuccess -> {
```
launch 블록은 제외해도 될 것 같은데, 지석님 의견은 어떠실까요?!
추가적으로 LaunchedEffect에서 uiEvent를 인자로 쓰면 어떤 일이 일어날까요?🤔

stopstone replied:
좋은 의견 감사합니다! 말씀하신 부 동의합니다!
확인해보니 이미 LaunchEffect 내부 람다에서 CoroutineScope를 제공하고 있어 불필요하네요!!

viewModel.uiEvent를 key로 사용하면, 아마 uiEvent 객체 자체를 key로 하기 때문에 viewmodel의 소멸 / 생성할 때 마다 LaunchEffect가 시작된다고 알고 있습니다.. 특별한 케이스에 대해서는 생각이 나는 것이 잘 없네요,,! 😢 고려해야할 부분이 있을까요/?

rhkrwngud445 replied:
제가 내부 값인 Event 타입과 헷갈렸군요😅 지석님께서 말씀주신대로 viewModel의 Flow 맵핑이 되어있어 실제 동작에서는 문제가 없는 것이 맞습니다! 
Flow가 불변이라서 해당 값으로 지정주신 것이면, Unit, true 등으로 최초 컴포지션 시 1회만 실행됨을 명시적으로 표현할 수 있을 것 같아요!

stopstone replied:
네! Unit으로 명확하게 해두는 것이 좋을 것 같네요! 확인해주셔서 감사합니다 :)

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/mothlyreminderall/MonthlyReminderAllViewModel.kt:
```diff
@@ -0,0 +1,133 @@
+            }.launchIn(viewModelScope)
+        }
+
+        private fun updateUIState() {
```
"UI 상태를 업데이트한다" 라는 의미를 가지기에는 추가적인 로직이 포함되어 있는 것 같습니다!
해당 함수명만 보고 내부 기능을 유추할 수 있게 수정해보면 어떨까요?
함수명이 길어도 함수 내부를 보는 것보다 이점이 있을 것 같아요!

stopstone replied:
명확하게 네이밍하는것이 중요한데 해당 부분에서 어떻게 네이밍을 해야할지 고민하다가 수정을 놓친 것 같습니다! 캐치 감사드립니다!

rhkrwngud445 reviewed (commented):
고생하셨습니다!
확장함수와 Preview 등 DX에 신경 많이 써주셨네요👍 감사합니다!
코멘트 조금 남겼는데, 확인 부탁드립니다!

stopstone committed:
refactor: MonthlyReminderAllScreen의 불필요한 launch 제거

`LaunchedEffect` 내에서 `collect`를 사용할 때 불필요한 `launch` 블록을 제거하여 코드를 간소화했습니다.

stopstone committed:
refactor: MonthlyReminderAllViewModel 내 불필요한 private backing property 제거

- `_monthlyReminders`와 `_completedReminders`에 대한 불필요한 `_` 접두사를 제거하여 코드를 간소화했습니다.

stopstone committed:
refactor: updateUIState 함수의 이름 변경

- `updateUIState` 함수의 역할을 더 명확하게 나타내기 위해 `combineRemindersToUIState`로 이름을 변경했습니다.
